### PR TITLE
MINIFICPP-1605 Always refresh AWS credentials through default credentials chain

### PR DESCRIPTION
**Pull request by lordgamez**
Due to a time skew problem when using temporary AWS credentials like instance profiles the credentials may expire on AWS side even though it can be non-expired when checked on client side. Because of this when using default credentials chain we should let the AWS SDK handle the caching and renewing the credentials and not cache manually on client side.

Jira ticket: https://issues.apache.org/jira/browse/MINIFICPP-1605

--------------------------------------------------------------------------------------------

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/test/aws-tests/AWSCredentialsServiceTest.cpp [hunk 1]
```diff
@@ -65,3 +77,28 @@ TEST_CASE_METHOD(AWSCredentialsServiceTestAccessor, "Test expired credentials ar
...
+  setEnvironmentCredentials("key2", "secret2");
+
+  // Check for credential refresh
+  REQUIRE(aws_credentials_impl->getAWSCredentials());
+  REQUIRE(aws_credentials_impl->getAWSCredentials()->GetAWSAccessKeyId() == "key2");
+  REQUIRE(aws_credentials_impl->getAWSCredentials()->GetAWSSecretKey() == "secret2");
+  REQUIRE_FALSE(aws_credentials_impl->getAWSCredentials()->IsExpired());
+}
+
+}
```
Comment (szaszm) [inline libminifi/test/aws-tests/AWSCredentialsServiceTest.cpp:104]:
```suggestion
}  // namespace
```

**Review by szaszm — APPROVED** [review]
LGTM after the linter error is fixed

File: libminifi/test/aws-tests/AWSCredentialsServiceTest.cpp (hunk 1 above)
Comment (lordgamez) [reply libminifi/test/aws-tests/AWSCredentialsServiceTest.cpp]: Fixed in c1a18bb5c2c60662e095ff43b63b2b0e53bd8635

**Review by arpadboda — APPROVED** [review]

**Comment by adamdebreceni** [comment]
as I see in `AWSCredentialsProvider::getAWSCredentials()` the `use_default_credentials_` branch falls through if we cannot acquire credentials from the default provider chain, is that intentional?

**Review by adamdebreceni — COMMENTED** [review]
as I see in `AWSCredentialsProvider::getAWSCredentials()` the `use_default_credentials_` branch falls through if we cannot acquire credentials from the default provider chain, is that intentional?

**Comment by lordgamez** [comment]
> as I see in `AWSCredentialsProvider::getAWSCredentials()` the `use_default_credentials_` branch falls through if we cannot acquire credentials from the default provider chain, is that intentional?

Yes, in that case we still try to check if we can find any other set credentials like the access key and secret key from properties or the credentials file as a backup.

**Comment by adamdebreceni** [comment]
> > as I see in `AWSCredentialsProvider::getAWSCredentials()` the `use_default_credentials_` branch falls through if we cannot acquire credentials from the default provider chain, is that intentional?
> 
> Yes, in that case we still try to check if we can find any other set credentials like the access key and secret key from properties or the credentials file as a backup.

and should we? I mean if I explicitly request the usage of default credentials shouldn't other user provided options be ignored?

**Comment by lordgamez** [comment]
> > > as I see in `AWSCredentialsProvider::getAWSCredentials()` the `use_default_credentials_` branch falls through if we cannot acquire credentials from the default provider chain, is that intentional?
> > 
> > 
> > Yes, in that case we still try to check if we can find any other set credentials like the access key and secret key from properties or the credentials file as a backup.
> 
> and should we? I mean if I explicitly request the usage of default credentials shouldn't other user provided options be ignored?

I see your point, but if the user explicitly sets the `Access Key` and the default credentials chain does not have any credentials to provide, should that be ignored as well? I think it could be a bigger problem for the user if they forget to turn off the default credential chain option and then after setting the credentials manually, they get an error that no credentials are available. It could take a long time to debug why those credentials were ignored.

I look at the `AWSCredentialsProvider` as more of a priority list, where if the default credential chain is set that should be the priority, but if that is not available then the other options should be checked as well.

**Review by adamdebreceni — APPROVED** [review]

**Pull request closed without merging**